L1Trigger/L1TMuon/src/GeometryTranslator.cc

```diff
@@ -0,0 +1,268 @@
+#include "L1Trigger/L1TMuon/interface/GeometryTranslator.h"
+#include "DataFormats/L1TMuon/interface/MuonTriggerPrimitive.h"
+
+// event setup stuff / geometries
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "Geometry/Records/interface/MuonGeometryRecord.h"
+
+#include "Geometry/CSCGeometry/interface/CSCGeometry.h"
+#include "L1Trigger/CSCCommonTrigger/interface/CSCConstants.h"
+#include "L1Trigger/CSCCommonTrigger/interface/CSCPatternLUT.h"
+
+#include "Geometry/DTGeometry/interface/DTGeometry.h"
+#include "L1Trigger/DTUtilities/interface/DTTrigGeom.h"
+#include "Geometry/RPCGeometry/interface/RPCGeometry.h"
+
+#include <cmath> // for pi
+
+using namespace l1t;
+
+GeometryTranslator::GeometryTranslator():  
+  _geom_cache_id(0ULL) {
+}
+
+GeometryTranslator::~GeometryTranslator() {  
+}
+
+double 
+GeometryTranslator::calculateGlobalEta(const MuonTriggerPrimitive& tp) const {
+  switch(tp.subsystem()) {
+  case MuonTriggerPrimitive::kDT:
+    return calcDTSpecificEta(tp);
+    break;
+  case MuonTriggerPrimitive::kCSC:
+    return calcCSCSpecificEta(tp);
+    break;
+  case MuonTriggerPrimitive::kRPC:
+    return calcRPCSpecificEta(tp);
+    break;
+  default:
+    return std::nan("Invalid TP type!"); 
+    break;
+  }
+}
+
+double 
+GeometryTranslator::calculateGlobalPhi(const MuonTriggerPrimitive& tp) const {
+  switch(tp.subsystem()) {
+  case MuonTriggerPrimitive::kDT:
+    return calcDTSpecificPhi(tp);
+    break;
+  case MuonTriggerPrimitive::kCSC:
+    return calcCSCSpecificPhi(tp);
+    break;
+  case MuonTriggerPrimitive::kRPC:
+    return calcRPCSpecificPhi(tp);
+    break;
+  default:
+    return std::nan("Invalid TP type!");
+    break;
+  }
+}
+
+double 
+GeometryTranslator::calculateBendAngle(const MuonTriggerPrimitive& tp) const {
+  switch(tp.subsystem()) {
+  case MuonTriggerPrimitive::kDT:
+    return calcDTSpecificBend(tp);
+    break;
+  case MuonTriggerPrimitive::kCSC:
+    return calcCSCSpecificBend(tp);
+    break;
+  case MuonTriggerPrimitive::kRPC:
+    return calcRPCSpecificBend(tp);
+    break;
+  default:
+    return std::nan("Invalid TP type!");
+    break;
+  }
+}
+
+void GeometryTranslator::checkAndUpdateGeometry(const edm::EventSetup& es) {
+  const MuonGeometryRecord& geom = es.get<MuonGeometryRecord>();
+  unsigned long long geomid = geom.cacheIdentifier();
+  if( _geom_cache_id != geomid ) {
```

Comment:
@davidlange6 It is a proper use of cacheIdentifier. However, I think the interface design makes it very easy to make a mistake. Essentially one must always be sure to call ```checkAndUpdateGeometry``` before ever calling any of the other methods. I would either pass the EventSetup in to all the calls or put the object into the EventSetup itself.